scripts/distribute_bpe_reward.py

```diff
@@ -0,0 +1,86 @@
+# coding: utf-8
+
+"""
+To distribute token-level rewards to sub-word units.
+"""
+
+import argparse
+import re
+import sys
+
+import numpy as np
+
+
+def main(args):
+
+    target_file = args.target
+    reward_file = args.reward
+
+    with open(target_file, 'r') as bpe_in, \
+         open(reward_file, 'r') as fbkin, open('feedback.bpe', 'w') as fbkout:
+
+        trans = bpe_in.readlines()
+        rewards = fbkin.readlines()
+
+        for idx, (sent, reward) in enumerate(zip(trans, rewards)):
+            reward_tok = reward.split()
+
+            if '@@' not in sent:
+                # No sub-words in this sentence
+                bpe_reward = reward_tok
+            else:
+                sent_str = sent.replace("@@ ", "")
+                sent_tok = sent.split()
+                assert(len(sent_str.split())==len(reward_tok)), \
+                    "input sentence length should be equal to number of reward"
+
+                if idx % args.print_interval == 0:
+                    print("sentence (post-process bpe):{}".format(
+                          sent_str[:-1]))
+                    print('input reward:{}'.format(" ".join(reward_tok)))
+
+                loc_bpe = [loc for loc, tok in enumerate(sent_tok) 
+                            if "@@" in tok]
+                loc_bpe_in_original = [bpeloc - len(loc_bpe[:idx]) 
+                                        for idx, bpeloc in enumerate(loc_bpe)]
+                loc_bpe_suffix = [loc + 1 for loc in loc_bpe]
+
+                bpe_reward = [None] * len(sent_tok)
+                bpe_count = 0
+                for ii in range(len(sent_tok)):
+                    if ii in loc_bpe_suffix:
+                        # the bpe suffix has the same reward as the bpe-prefix
+                        orig_loc_bpe_prefix = loc_bpe_in_original[loc_bpe_suffix.index(ii)]
+                        bpe_reward[ii] = reward_tok[orig_loc_bpe_prefix]
+                        if idx % args.print_interval == 0:
+                            print('BPE suffix! loc:{} ; token[loc]:{}; orig_loc:{}; '
+                                  'reward[orig_loc]:{}'.format(ii, sent_tok[ii], 
+                                  orig_loc_bpe_prefix, reward_tok[orig_loc_bpe_prefix]))
+                        bpe_count += 1
+                    else:
+                        bpe_reward[ii] = reward_tok[ii-bpe_count]
+
```

Comment:
maybe assert bpe_count == len(loc_bpe)-1, just to be sure

Reply:
bpe_count is a variable for keep tracking the number of bpe founds while looping through the whole tokenised sentence. 

I add an assertion at the end of the loop to ensure that the number of bpe found eventually is the same as the length of of the list holding locations of bpe_prefix and bpe_suffix